File: android_flutter_claude/transcripto/packages/domain/lib/src/use_cases/encrypt_text.dart
import 'package:core_common/core_common.dart';
import '../entities/cipher_method.dart';
import '../entities/cipher_params.dart';
import '../entities/envelope.dart';
import '../repositories/cipher_repository.dart';

/// Use case for encrypting text
class EncryptText {
  const EncryptText(this._cipherRepository);
  
  final CipherRepository _cipherRepository;
  
  /// Encrypts the input text and optionally wraps in envelope format
  Result<String> call(String input, CipherParams params, {bool useEnvelope = true}) {
    try {
      // Validate input
      if (input.trim().isEmpty) {
        return const Failure(ValidationException('El texto de entrada no puede estar vacío'));
      }
      
      // Validate parameters based on method
      final validation = _validateParams(params);
      if (validation != null) {
        return Failure(ValidationException(validation));
      }
      
      // Perform encryption
      final result = _cipherRepository.encrypt(input, params);
      
      return result.fold(
        (encryptedText) {
          if (useEnvelope) {
            final envelope = Envelope.fromParams(params, encryptedText);
            return Success(envelope.serialize());
          } else {
            return Success(encryptedText);
          }
        },
        (error) => Failure(error),
      );
    } catch (e, stackTrace) {
      return Failure(CipherException('Error durante el cifrado: $e'), stackTrace);
    }
  }
  
  String? _validateParams(CipherParams params) {
    switch (params.method) {
      case CipherMethod.caesar:
        if (params.shift == null) {
          return 'El desplazamiento es requerido para César';
        }
        if (params.shift! < -25 || params.shift! > 25) {
          return 'El desplazamiento debe estar entre -25 y 25';
        }
        break;
      case CipherMethod.vigenere:
        if (params.key == null || params.key!.trim().isEmpty) {
          return 'La clave es requerida para Vigenère';
        }
        break;
      case CipherMethod.xor:
        if (params.key == null || params.key!.trim().isEmpty) {
          return 'La clave es requerida para XOR';
        }
        break;
      case CipherMethod.base64:
        // No additional validation needed
        break;
    }
    
    return null;
  }
}

File: android_kimi/domain/src/main/java/com/transcripto/domain/usecase/EncryptText.kt
package com.transcripto.domain.usecase

import com.transcripto.domain.repository.CipherMethod
import com.transcripto.domain.model.CipherParams
import com.transcripto.domain.model.CipherResult
import com.transcripto.domain.repository.CipherRepository
import javax.inject.Inject

/**
 * Use case for encrypting text using various cipher methods.
 */
class EncryptText @Inject constructor(
    private val cipherRepository: CipherRepository
) {
    /**
     * Encrypts the input text using the specified method and parameters.
     * @param input The text to encrypt
     * @param params The encryption parameters
     * @return CipherResult containing encrypted text or error
     */
    operator fun invoke(input: String, params: CipherParams): CipherResult {
        if (input.isBlank()) {
            return CipherResult.Error("El texto de entrada no puede estar vacío")
        }

        val cipherMethod = cipherRepository.getCipher(params.method)
            ?: return CipherResult.Error("Método de cifrado no soportado")

        return cipherMethod.encrypt(input, params)
    }
}
